pkg/apis/allocation/types.go

```diff
@@ -0,0 +1,121 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package allocation
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IPRange defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type IPRange struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPRangeSpec   `json:"spec,omitempty"`
+	Status IPRangeStatus `json:"status,omitempty"`
+}
+
+// IPRangeSpec describe how the IPRange's specification looks like.
+type IPRangeSpec struct {
+	// Range of IPs in CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+	Range string `json:"range"`
+	// Primary indicates if this is the primary allocator to be used by the
+	// apiserver to allocate IP addresses.
+	// NOTE this can simplify the Service strategy logic so we don't have to infer
+	// the primary allocator, it also may allow to switch between primary families in
+	// a cluster, but this looks like a loooong shot.
+	// +optional
+	Primary bool `json:"primary,omitempty"`
+}
+
+// IPRangeStatus defines the observed state of IPRange.
+type IPRangeStatus struct {
+	// Free represent the number of IP addresses that are not allocated in the Range.
+	// +optional
+	Free int64 `json:"free,omitempty"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IPRangeList contains a list of IPRange objects.
+type IPRangeList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []IPRange `json:"items"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// IPAddress represents an IP used by Kubernetes associated to an IPRange.
+// The name of the object is the IP address decimal number, because colons
+// are not allowed and IPv6 addresses have different text representations.
+// xref: https://tools.ietf.org/html/rfc4291
+type IPAddress struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPAddressSpec   `json:"spec,omitempty"`
+	Status IPAddressStatus `json:"status,omitempty"`
+}
+
+// IPRangeRef contains information that points to the IPRange being used
+type IPRangeRef struct {
+	// APIGroup is the group for the resource being referenced
+	APIGroup string
+	// Kind is the type of resource being referenced
+	Kind string
+	// Name is the name of resource being referenced
+	Name string
+}
+
+// IPAddressSpec describe the attributes in an IP Address,
+type IPAddressSpec struct {
+	// Address is the text representation of the IP Address.
+	Address string `json:"address"`
+	// IPRangeRef references the IPRange associated to this IP Address.
+	IPRangeRef IPRangeRef `json:"ipRangeRef,omitempty"`
```

Comment:
Can't we figure out the range from looking at the IP address?

Reply:
yeah ... but then you can only have one service subnet per IP family ... and this can be used to detect orphan resources and to disallow creating IPAddresses not associated to a valid IPRange

Reply:
I guess I was trying to optimize space, but if this is 1 IP for service, maybe it's not that important for this object to be small. If this was one per pod I don't think we could do this.

Reply:
This is for services only, ... Pods are assigned by the CNI plugin, somehow, I;d say they are not really kubernetes resources